controllers/machine_controller_test.go

```diff
@@ -1856,6 +1860,193 @@ func TestNodeToMachine(t *testing.T) {
+		},
+		Status: clusterv1.MachineStatus{
+			NodeRef: &corev1.ObjectReference{
+				Name: "test",
+			},
+		},
+	}
+
+	cpmachine1 := &clusterv1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "cp1",
+			Namespace: metav1.NamespaceDefault,
+			Labels: map[string]string{
+				clusterv1.ClusterLabelName:             "test-cluster",
+				clusterv1.MachineControlPlaneLabelName: "",
+			},
+			Finalizers: []string{clusterv1.MachineFinalizer},
+		},
+		Spec: clusterv1.MachineSpec{
+			ClusterName:       "test-cluster",
+			InfrastructureRef: corev1.ObjectReference{},
+			Bootstrap:         clusterv1.Bootstrap{DataSecretName: pointer.StringPtr("data")},
+		},
+		Status: clusterv1.MachineStatus{
+			NodeRef: &corev1.ObjectReference{
+				Name: "cp1",
+			},
+		},
+	}
+
+	testCases := []struct {
+		name               string
+		deletionTimeout    *metav1.Duration
+		resultErr          bool
+		clusterDeleted     bool
+		expectNodeDeletion bool
+		createFakeClient   func(...client.Object) client.Client
+	}{
+		{
+			name:               "should return no error when deletion is successful",
+			deletionTimeout:    &metav1.Duration{Duration: time.Second},
+			resultErr:          false,
+			expectNodeDeletion: true,
+			createFakeClient: func(initObjs ...client.Object) client.Client {
+				return fake.NewClientBuilder().
+					WithObjects(initObjs...).
+					Build()
+			},
+		},
+		{
+			name:               "should return an error when timeout is not expired and node deletion fails",
+			deletionTimeout:    &metav1.Duration{Duration: time.Hour},
+			resultErr:          true,
+			expectNodeDeletion: false,
+			createFakeClient: func(initObjs ...client.Object) client.Client {
+				fc := fake.NewClientBuilder().
+					WithObjects(initObjs...).
+					Build()
+				return fakeClientWithNodeDeletionErr{fc}
+			},
+		},
+		{
+			name:               "should return an error when timeout is infinite and node deletion fails",
+			deletionTimeout:    &metav1.Duration{Duration: 0}, // should lead to infinite timeout
+			resultErr:          true,
+			expectNodeDeletion: false,
+			createFakeClient: func(initObjs ...client.Object) client.Client {
+				fc := fake.NewClientBuilder().
+					WithObjects(initObjs...).
+					Build()
+				return fakeClientWithNodeDeletionErr{fc}
+			},
+		},
+		{
+			name:               "should not return an error when timeout is expired and node deletion fails",
+			deletionTimeout:    &metav1.Duration{Duration: time.Millisecond},
+			resultErr:          false,
+			expectNodeDeletion: false,
+			createFakeClient: func(initObjs ...client.Object) client.Client {
+				fc := fake.NewClientBuilder().
+					WithObjects(initObjs...).
+					Build()
+				return fakeClientWithNodeDeletionErr{fc}
+			},
+		},
+		{
+			name:               "should not delete the node or return an error when the cluster is marked for deletion",
+			deletionTimeout:    nil, // should lead to infinite timeout
```

Comment:
nit: drop misleading comment

Reply:
It actually behaves like that. It should never occur because of the validating webhook, but if the timeout is actually nil it will retry indefinitely.